State: Type route path & query

chriskmnds opened the pull request:
#### Changes proposed in this Pull Request

Adds types to the route path and query state actions/reducers/selectors.

I guess more typing can't really hurt. This will also help our work in #55313 with better types coming from the underlying path & query selectors. Hopefully, put some of the groundwork here.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* TBD

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #55313, #57109

chriskmnds committed:
Type current path & query selectors

chriskmnds commented on client/state/route/types.ts:
```diff
@@ -0,0 +1,45 @@
+ */
+import { ROUTE_CLEAR_LAST_NON_EDITOR, ROUTE_SET } from 'calypso/state/action-types';
+
+export type Query = {
+	[ key: string ]: string | number | undefined;
+	_timestamp?: number;
+};
```
Not fun of this type. Hopefully, there's a way to express it differently. `_timestamp` is the only property of type `number` that may also be undefined - everything else is a `string`.

jsnajdr replied:
The `_timestamp` shouldn't really be here on the `query`, but rather on a separate property of the `state.route` object. There would be `{ path, query, timestamp }` properties there.

The `_timestamp` field is currently used only by the Guided Tours code and nowhere else. The meaning is "the time at which the last (initial, previous) route navigation happened".

Adding types reveals such hacks, like slapping non-query data on the `query` object, and nudges us toward fixing them.

jsnajdr commented on client/state/route/types.ts:
```diff
@@ -0,0 +1,45 @@
+
+export type RouteSetActionCreator = (
+	path: RouteSetAction[ 'path' ],
+	query: RouteSetAction[ 'query' ]
```
The types here are `Path` and `Query`, there's no need to specify them in a verbose way like this.

jsnajdr commented on client/state/route/query/reducer.ts:
```diff
@@ -0,0 +1,46 @@
+const currentReducer = ( state: QueryState[ 'current' ], query: Query ) =>
+	! isEqualQuery( state, query ) ? timestamped( query ) : state;
+
+const initialState = {
```
```ts
const initialState: QueryState = {
  initial: false,
  ...
}
```
is better. It avoids the need to write `as const`.

jsnajdr commented on client/state/route/types.ts:
```diff
@@ -0,0 +1,45 @@
+	previous: Query | false;
+};
+
+export type Path = string;
```
Type aliases like this are not very useful in TypeScript, where the type checking is structural rather than nominal. Simply using `string` everywhere is OK.

jsnajdr commented on client/state/route/types.ts:
```diff
@@ -0,0 +1,45 @@
+
+export type RouteState = {
+	path: PathState;
+	query: QueryState;
```
There is also a `lastNonEditorRoute: string` field in `RouteState`

jsnajdr commented on client/state/selectors/get-current-query-arguments.ts:
```diff
@@ -0,0 +1,26 @@
+ */
+export function getCurrentQueryArguments(
+	state: IAppState
+): IAppState[ 'route' ][ 'query' ][ 'current' ] | null {
```
The return type is `Query`, so why not use it instead?

jsnajdr commented on client/state/route/last-non-editor-route/reducer.ts:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Internal dependencies
+ */
+import { ROUTE_CLEAR_LAST_NON_EDITOR, ROUTE_SET } from 'calypso/state/action-types';
```
This PR adds a `reducer.ts` file, but doesn't delete the old `reducer.js` one.

jsnajdr commented on client/state/route/types.ts:
```diff
@@ -0,0 +1,45 @@
+	path: Path;
+	query: Query;
+	type: typeof ROUTE_SET;
+};
```
I think that diligently declaring strict types for the `Route*` actions is not necessary. The types are not actually used anywhere.

The `setRoute` action creator (like any other action creator) returns an opaque action object whose internal structure is not interesting. The action is only dispatched by passing it to `dispatch` and its fields could as well be private.

It's only the reducer that actually reads them, but our reducer is declared as accepting `AnyAction` and doesn't use the specific types.

Because of that, the actions creators could simply return the `Action` type from `redux`.

Commit:
log-in/new without query parameters threw error

getCurrentQueryArguments was used, accessing email_address, without checking if it was null

Commit:
Fixed error in useHomeLayoutQueryParams

Removed unused types

jsnajdr commented on client/data/home/use-home-layout-query-params.ts:
```diff
@@ -8,7 +9,8 @@ export interface HomeLayoutQueryParams {
 export function useHomeLayoutQueryParams(): HomeLayoutQueryParams {
-	const { dev, view } = useSelector( getCurrentQueryArguments ) as any;
+	const dev = get( useSelector( getCurrentQueryArguments ), 'dev', undefined );
+	const view = get( useSelector( getCurrentQueryArguments ), 'view', undefined );
```
This doesn't look like an improvement. Just `const { dev, view } = useSelector( ... )` should work. If it doesn't, there's something wrong with the types and they should be fixed. Calling the selector twice and introducing a new usage of Lodash is not good.

jsnajdr commented on client/state/selectors/get-current-query-arguments.ts:
```diff
@@ -0,0 +1,25 @@
+ * Kept for backwards compatibility in case any null checks.
+ * Investigate and refactor.
+ */
+export function getCurrentQueryArguments( state: IAppState ): Query | false | null {
```
The return type should be just `Query`, and the selector should return an empty `{}` object if the current URL doesn't have a query string. That's the API most convenient for the caller.

chriskmnds replied:
I feel the selector could just return whatever `state.route.query.current` points to. 🤔 No need to coalesce `null` in the mix as we do here either (as also done previously). So either `undefined` or `state.route.query.current` -> ultimately, just `return state?.route?.query?.current;`. 

This makes sense to me in the following way: 
- the selector becomes a simple/dumb reflection of the state tree
- the default `{}` (or whatever else we choose) becomes the responsibility of the reducer
- undefined for something that has not been defined yet may carry meaning of its own, and a raw/base selector can reflect that

Of course, trade-offs e.g. a derived state selector will ultimately provide some sort of default, which they'd be responsible for defining anyhow. So the alternative (what you mention I think) might just be better e.g. selectors always defining and returning the default values.

jsnajdr replied:
Yes, I think the `false` value for the state doesn't need to be there, in neither of `initial`, `current` and `previous`. It can be an empty object instead:
```js
const EMPTY_QUERY = {};

const initialState = {
  initial: EMPTY_QUERY,
  current: EMPTY_QUERY,
  previous: EMPTY_QUERY,
};
```
And the reducer for `initial` can do a check like:
```js
initial = state.initial === EMPTY_QUERY ? action.query : state.initial
```
to set the `initial` state only on the very first `ROUTE_SET`. Even if `action.query` will be an empty object, it will be a _different_ empty object than the initial state.

chriskmnds replied:
That's perfect. Thanks! We'll take this path.

chriskmnds commented on client/state/selectors/get-current-query-arguments.ts:
```diff
@@ -17,9 +18,7 @@ import type { IAppState } from 'calypso/state/types';
-	state: IAppState
-): IAppState[ 'route' ][ 'query' ][ 'current' ] | null {
+export function getCurrentQueryArguments( state: IAppState ): Query | false | null {
 	return state?.route?.query?.current || null;
```
I think the selector should just return whatever `state.route.query.current` points to. No need to coalesce `null` in the mix as we do here either (as also done previously). So either `undefined` or `state.route.query.current` -> ultimately, just `return state?.route?.query:?.current;`.
